### PR TITLE
add functionality to clone graphs (important for unit test performance)

### PR DESCRIPTION
**Pull request by mpollmeier**
Alternative proposal to https://github.com/joernio/flatgraph/pull/162
I moved a few things around and refactored for readability. Main difference in API is that rather than
`graph.copyDataFrom(other: Graph)` 
this does
`graph.copy(): Graph`

### CODE REVIEW

**Review by bbrehm — APPROVED** [review]
I'm fine with this API -- if you prefer then we go with this.

You still need to include https://github.com/joernio/flatgraph/pull/162/commits/37514cbd64066bdf59b7581ccfa8e0707ce7bf89 , though -- my original commit was plain incorrect, as your test nicely uncovered, and needs copy-on-write for edge-properties in the diffgraph applier.

File: core/src/main/scala/flatgraph/Graph.scala [hunk 1]
```diff
@@ -143,4 +144,33 @@ class Graph(val schema: Schema, val storagePathMaybe: Option[Path] = None) exten
     }
   }
 
+  def cloneDataFrom(other: Graph): this.type = {
```
Comment (bbrehm) [inline core/src/main/scala/flatgraph/Graph.scala:147]: I think you want to remove this? (only one API, and copy is fine)

Comment (mpollmeier) [reply core/src/main/scala/flatgraph/Graph.scala:147]: oh, yes of course, thanks for catching :)

**Comment by mpollmeier** [comment]
> You still need to include https://github.com/joernio/flatgraph/commit/37514cbd64066bdf59b7581ccfa8e0707ce7bf89 

The test is green, so technically no, because I added a `.clone` in `cloneThing`: https://github.com/joernio/flatgraph/pull/163/files#diff-c01ea7a83957bac3955efb6f2fb9379aa20ab5ff53d2fccd82e41dd57b7633d1R50

But I can (also)? add it to the DiffGraphApplier if you prefer..?

**Comment by mpollmeier** [comment]
rebased, cleaned, formatted, included your commit

looking good?

File: core/src/main/scala/flatgraph/DiffGraphApplier.scala [hunk 2]
```diff
@@ -408,7 +408,7 @@ private[flatgraph] class DiffGraphApplier(graph: Graph, diff: DiffGraphBuilder)
     val oldQty = graph.neighbors(pos).asInstanceOf[Array[Int]]
     val edgeProp = graph.neighbors(pos + 2) match {
       case _: DefaultValue => graph.schema.allocateEdgeProperty(nodeKind, direction, edgeKind, size)
-      case other           => other
+      case other: Array[_] => other.clone() // this is not strictly necessary, but everything else is copy-on-write
```
Comment (bbrehm) [inline core/src/main/scala/flatgraph/DiffGraphApplier.scala:411]: I think we should remove this copy and replace it with a comment that we will need to update the code once we switch away from full copy-on-write.

If we're eating the extra allocs for graph updates, can as well reap the benefits in terms of faster clone.

**Comment by bbrehm** [comment]
I'm fine, you can merge and close the superseded sibling.

I would remove the defensive copy of arrays in the `copy` -- but we can also leave it in if you prefer, copying is only supposed to happen for pretty small graphs in unit tests anyways.

File: core/src/main/scala/flatgraph/DiffGraphApplier.scala (hunk 2 above)
Comment (mpollmeier) [reply core/src/main/scala/flatgraph/DiffGraphApplier.scala]: yup that's what I meant with my comment in https://github.com/joernio/flatgraph/pull/163#issuecomment-2016856624

I removed the clone again and added a comment.

**Pull request merged**